src/main/python/fine_tuning/x_fold_cv.py

```diff
@@ -0,0 +1,121 @@
+                    if basemodel not in avg_performances:
+                        avg_performances[basemodel] = {}
+                    if model not in avg_performances[basemodel]:
+                        avg_performances[basemodel][model] = {}
+
+                    param_avg_performances = self._get_param_avg_performances(os.path.join(eval_dir,fn))
+                    for metric in param_avg_performances:
+                        if metric not in avg_performances[basemodel][model]:
+                            avg_performances[basemodel][model][metric] = {}
+                        for i in param_avg_performances[metric]:
+                            if i not in avg_performances[basemodel][model][metric]:
+                                avg_performances[basemodel][model][metric][i] = {}
+                            avg_performances[basemodel][model][metric][i][param] = param_avg_performances[metric][i]
+
+        return avg_performances
+
+    def tune(self,quite):
+        # x-fold tuning
+        avg_performances = self._get_param_average()
+        for basemodel in avg_performances:
+            for model in avg_performances[basemodel]:
+                for metric in avg_performances[basemodel][model]:
+                    print "For %s %s %s" %(basemodel, model, metric)
+                    metric_fold_performances = []
+                    for i in xrange(self.fold):
+                        test_fold_performances = avg_performances[basemodel][model][metric][i]
+                        training_data = {}
+                        for j in xrange(self.fold):
+                            if j == i:
+                                continue
+                            fold_performance = avg_performances[basemodel][model][metric][j]
+                            for param in  fold_performance:
+                                if param not in training_data:
+                                    training_data[param] = .0
+                                training_data[param] += fold_performance[param]
+                        sorted_training_performance = sorted(training_data.items(),
+                                                             key=lambda x:x[1],
+                                                             reverse=True)
+                        best_param = sorted_training_performance[0][0]
+                        if not quite:
+                            print "\tFold: %d" %(i)
+                            print "\t\tBest param: %s" %(best_param)
+                            print "\t\ttest performance: {0:.4f}".format(test_fold_performances[best_param])
+
+                        metric_fold_performances.append(test_fold_performances[best_param])
+                    print "\tAverage {0:.4f}".format( sum(metric_fold_performances) / len(metric_fold_performances))
+
+    def _get_param_avg_performances(self,file_path):
+        param_performance_list = {}
+        for i in xrange(self.fold):
```

Comment:
replace `i` with fold across the file

Reply:
`fold_idx` maybe? We already have `self.fold` ...